---
title: index
weight: 10
---

```plain
$ lexicmap index -h
Generate an index from FASTA/Q sequences

Input:
 *1. Sequences of each reference genome should be saved in separate FASTA/Q files, with reference identifiers
     in the file names.
  2. Input plain or gzip/xz/zstd/bzip2 compressed FASTA/Q files can be given via positional arguments or
     the flag -X/--infile-list with a list of input files.
     Flag -S/--skip-file-check is optional for skipping file checking if you trust the file list.
  3. Input can also be a directory containing sequence files via the flag -I/--in-dir, with multiple-level
     sub-directories allowed. A regular expression for matching sequencing files is available via the flag
     -r/--file-regexp.
  4. Some non-isolate assemblies might have extremely large genomes (e.g., GCA_000765055.1, >150 mb).
     The flag -g/--max-genome is used to skip these input files, and the file list would be written to a file
     (-G/--big-genomes).
     You need to increase the value for indexing fungi genomes.
  5. Maximum genome size: 268,435,456.
     More precisely: $total_bases + ($num_contigs - 1) * 1000 <= 268,435,456, as we insert 1000-bp intervals
     of N's between contigs to reduce the sequence scale to index.

  Attention:
   *1) ► You can rename the sequence files for convenience, e.g., GCF_000017205.1.fa.gz, because the genome
       identifiers in the index and search result would be: the basenames of files with common FASTA/Q file
       extensions removed, which are extracted via the flag -N/--ref-name-regexp.
       ► The extracted genome identifiers better be distinct, which will be shown in search results
       and are used to extract subsequences in the command "lexicmap utils subseq".
    2) ► Unwanted sequences like plasmids can be filtered out by content in FASTA/Q header via regular
       expressions (-B/--seq-name-filter).
    3) All degenerate bases are converted to their lexicographic first bases. E.g., N is converted to A.
        code  bases    saved
        A     A        A
        C     C        C
        G     G        G
        T/U   T        T

        M     A/C      A
        R     A/G      A
        W     A/T      A
        S     C/G      C
        Y     C/T      C
        K     G/T      G

        V     A/C/G    A
        H     A/C/T    A
        D     A/G/T    A
        B     C/G/T    C

        N     A/C/G/T  A

Important parameters:

  --- Genome data ---
 *1. -b/--batch-size,       ► Maximum number of genomes in each batch (maximum: 131072, default: 5000).
                            ► If the number of input files exceeds this number, input files are split into multiple
                            batches and indexes are built for all batches. In the end, seed files are merged, while
                            genome data files are kept unchanged and collected.
                            ■ Bigger values increase indexing memory occupation and increase batch searching speed,
                            while single query searching speed is not affected.

  --- LexicHash mask generation ---
  0. -M/--mask-file,        ► File with custom masks, which could be exported from an existing index or newly
                            generated by "lexicmap utils masks".
                            This flag oversides -k/--kmer, -m/--masks, -s/--rand-seed, etc.
 *1. -k/--kmer,             ► K-mer size (maximum: 32, default: 31).
                            ■ Bigger values improve the search specificity and do not increase the index size.
 *2. -m/--masks,            ► Number of LexicHash masks (default: 40000).
                            ■ Bigger values improve the search sensitivity, increase the index size, and slow down
                            the search speed.
  3. -p/--seed-min-prefix,  ► Minimum length of shared substrings (anchors) in searching (maximum: 32, default: 15).
                            ► This value is used to remove masks with a prefix of low-complexity.

  --- Seeds data (k-mer-value data) ---
 *1. --seed-max-desert      ► Maximum length of distances between seeds (default: 200).
                            The default value of 200 guarantees queries >=200 bp would match at least one seed.
                            ► Large regions with no seeds are called sketching deserts. Deserts with seed distance
                            larger than this value will be filled by choosing k-mers roughly every
                            --seed-in-desert-dist (50 by default) bases.
                            ■ Big values decrease the search sensitivity for distant targets, speed up the indexing
                            speed, decrease the indexing memory occupation and decrease the index size. While the
                            alignment speed is almost not affected.
  2. -c/--chunks,           ► Number of seed file chunks (maximum: 128, default: #CPUs).
                            ► Bigger values accelerate the search speed at the cost of a high disk reading load.
                            The maximum number should not exceed the maximum number of open files set by the
                            operating systems.
 *3. -J/--seed-data-threads ► Number of threads for writing seed data and merging seed chunks from all batches
                            (maximum: -c/--chunks, default: 8).
                            ■ Bigger values increase indexing speed at the cost of slightly higher memory occupation.
  4. --partitions,          ► Number of partitions for indexing each seed file (default: 512).
                            ► Bigger values bring a little higher memory occupation. 512 is a good value with high
                            searching speed, Larger or smaller values would decrease the speed in "lexicmap search".
                            ► After indexing, "lexicmap utils reindex-seeds" can be used to reindex the seeds data
                            with another value of this flag.
  5. --max-open-files,      ► Maximum number of open files (default: 512).
                            ► It's only used in merging indexes of multiple genome batches.

Usage:
  lexicmap index [flags] [-k <k>] [-m <masks>] { -I <seqs dir> | -X <file list>} -O <out dir>

Flags:
  -b, --batch-size int            ► Maximum number of genomes in each batch (maximum value: 131072)
                                  (default 5000)
  -G, --big-genomes string        ► Out file of skipped files with genomes >= -g/--max-genome
  -c, --chunks int                ► Number of chunks for storing seeds (k-mer-value data) files.
                                  (default 16)
  -r, --file-regexp string        ► Regular expression for matching sequence files in -I/--in-dir,
                                  case ignored. (default
                                  "\\.(f[aq](st[aq])?|fna)(\\.gz|\\.xz|\\.zst|\\.bz2)?$")
      --force                     ► Overwrite existing output directory.
  -h, --help                      help for index
  -I, --in-dir string             ► Input directory containing FASTA/Q files. Directory and file
                                  symlinks are followed.
  -k, --kmer int                  ► Maximum k-mer size. K needs to be <= 32. (default 31)
  -M, --mask-file string          ► File of custom masks. This flag oversides -k/--kmer, -m/--masks,
                                  -s/--rand-seed, -p/--seed-min-prefix, etc.
  -m, --masks int                 ► Number of LexicHash masks. (default 40000)
  -g, --max-genome int            ► Maximum genome size. Extremely large genomes (e.g., non-isolate
                                  assemblies from Genbank) will be skipped. Need to be smaller than the
                                  maximum supported genome size: 268435456 (default 15000000)
      --max-open-files int        ► Maximum opened files, used in merging indexes. (default 512)
  -O, --out-dir string            ► Output LexicMap index directory.
      --partitions int            ► Number of partitions for indexing seeds (k-mer-value data) files.
                                  (default 512)
  -s, --rand-seed int             ► Rand seed for generating random masks. (default 1)
  -N, --ref-name-regexp string    ► Regular expression (must contains "(" and ")") for extracting the
                                  reference name from the filename. (default
                                  "(?i)(.+)\\.(f[aq](st[aq])?|fna)(\\.gz|\\.xz|\\.zst|\\.bz2)?$")
      --save-seed-pos             ► Save seed positions, which can be inspected with "lexicmap utils
                                  seed-pos".
  -J, --seed-data-threads int     ► Number of threads for writing seed data and merging seed chunks
                                  from all batches, the value should be in range of [1, -c/--chunks]
                                  (default 8)
  -d, --seed-in-desert-dist int   ► Distance of k-mers to fill deserts. (default 50)
  -D, --seed-max-desert int       ► Maximum length of sketching deserts, or maximum seed distance.
                                  Deserts with seed distance larger than this value will be filled by
                                  choosing k-mers roughly every --seed-in-desert-dist bases. (default 200)
  -p, --seed-min-prefix int       ► Minimum length of shared substrings (anchors) in searching. Here,
                                  this value is used to remove low-complexity masks and choose k-mers to
                                  fill sketching deserts. (default 15)
  -B, --seq-name-filter strings   ► List of regular expressions for filtering out sequences by
                                  contents in FASTA/Q header/name, case ignored.
  -S, --skip-file-check           ► Skip input file checking when given files or a file list.

Global Flags:
  -X, --infile-list string   ► File of input file list (one file per line). If given, they are
                             appended to files from CLI arguments.
      --log string           ► Log file.
      --quiet                ► Do not print any verbose information. But you can write them to a file
                             with --log.
  -j, --threads int          ► Number of CPU cores to use. By default, it uses all available cores.
                             (default 16)
```


## Examples

See {{< button size="small" relref="tutorials/index" >}}Building an index{{< /button >}}
